src/plugin/advancedCompare/index.js

```diff
@@ -0,0 +1,43 @@
+import * as C from '../../constant'
+
+export default (option, dayjsClass) => {
+  const proto = dayjsClass.prototype
+  const oldSame = proto.isSame
+  const oldIsAfter = proto.isAfter
+  const oldIsBefore = proto.isBefore
+  const pretty = proto.$utils().prettyUnit
+
+  const v = (l, r) => l.isValid() && r.isValid()
+
+  proto.isSame = function (o, u) {
+    if (!v(this, o)) return false
+
+    return pretty(u) === C.MS
+      ? oldSame.bind(this)(o)
+      : this.startOf(u).valueOf() <= o && o <= this.endOf(u).valueOf()
+  }
+
+  proto.isAfter = function (o, u) {
+    if (!v(this, o)) return false
+
+    return pretty(u) === C.MS
+      ? oldIsAfter.bind(this)(o)
+      : o.valueOf() < this.startOf(u).valueOf()
+  }
+
+  proto.isBefore = function (o, u) {
+    if (!v(this, o)) return false
+
+    return pretty(u) === C.MS
```

Comment:
seems we could compare this directly, just `return this.endOf(u).valueOf() < o.valueOf()`, since 'ms' is the default value of `endOf` and `startOf`

Reply:
Good point, that leaves us with just

```javascript
proto.isSame = function (o, u) {
    return v(this, o) && this.startOf(u).valueOf() <= o && o <= this.endOf(u).valueOf()
  }

  proto.isAfter = function (o, u) {
    return v(this, o) && o.valueOf() < this.startOf(u).valueOf()
  }

  proto.isBefore = function (o, u) {
    return v(this, o) && this.endOf(u).valueOf() < o.valueOf()
  }
```

Tests are green, should I push that?

Reply:
Integrated your suggestions with a101e3e437724f7498c8f651fd90061c35606389.

Reply:
Looks great. 

We might could modify `isSame`  `isAfter ` and `isBefore` in our main bundle instead of a separate plugin. 

And leave the other two as two individual plugins like `isBetween` plugin. 

How do you think?

Reply:
Sounds great to me. I wasn't sure if you consider these comparisons to be important enough to be in the main bundle, but given how you helped me reduce the implementation to just oneliners, I think this brings great benefits with relatively low costs in terms of size.

What would be a proper plugin name for the remaining two methods `isSameOrAfter()` and `isSameOrBefore()`? Do you want me to do the changes or will you do them?

Reply:
If we skip the validation check, seems it won't increase bundle size too much (or the same maybe). 

Just have a try if you have spare time, and let us know the size of the modified main bundle.

We better use API name as its plugin name as well as file name.

Reply:
Yes, and two separate plugins might be a good choice I think

Reply:
Alright, so I am going to

- swap `isBefore(other)`, `isAfter(other)` and `isSame(other)` with `isBefore(other, unit?)`, `isAfter(other, unit?)` and `isSame(other, unit?)` in the main bundle
- integrate the test suite into the main bundle test suite (should it be in a separate file? it is a rather large body of tests on its own...)
- add a plugin `isSameOrBefore` named `is-same-or-before.js`
- add a plugin `isSameOrAfter` named `is-same-or-after.js`

Agree?

Reply:
Agreed,

I like your large test suite, and better keep it as a separate file.

Plugin file name use camel-like case like the other.

Plus, we might need to update `isBetween` plugin as well.

Also our `index.d.ts` file.

Reply:
- [ ] Move `isSame(other, units)`, `isAfter(other, units)`, `isBefore(other, units)` to main bundle
- [ ] Adjust tests to run against the main bundle
- [ ] Move `isSameOrAfter(other, units)` to `plugin/isSameOrAfter.js`
- [ ] Move `isSameOrBefore(other, units)` to `plugin/isSameOBefore.js`
- [ ] Adjust `isBetween()` to also support units
- [ ] Adjust Typescript typings in `index.d.ts` to reflect the changes

Reply:
Apart from `isBetween`, all tasks are done as of c13c93d280f24c970db5e72752156fc1a4d5e7b2. Please check that I did not miss anything with Typescript typings, not too sure about that one, especially the tests for it.

Size of the main bundle is now:

```bash
> size-limit && gzip-size dayjs.min.js

  Package size: 2.37 KB
  Size limit:   2.99 KB
  With all dependencies, minified and gzipped

2.57 kB
```

Reply:
`isBetween()` added with d0eeb32. Please do proper review, those are bigger changes. If no objections from your side, that should be it.